src/sensitivities/chainrules.jl

```diff
@@ -0,0 +1,141 @@
+function generate_overload(sig)
+    opT, argTs = Iterators.peel(ExprTools.parameters(sig))
+    opT <: Core.Builtin && return nothing  # can't do operater overloading for builtins
+    opT <: Function || return nothing  # not handling non-functions
+    fieldcount(opT) == 0 || return nothing # not handling functors
+
+    nameof(opT.name.module) == :NaNMath  && return  # Don't care about NaNMath
+
+    # Ignore functions that have complex ranges. This may change when Nabla supports complex
+    # numbers.
+    opT ∈ typeof.((
+        SpecialFunctions.hankelh1, SpecialFunctions.hankelh2,
+        log1p, rem2pi, mod, atan, rem,
+    ))
+    isempty(argTs) && return  # we are an operator overloading AD, need operands
+    signature_def = build_def(sig)
+    original_signature_args = signature_def[:args]
+    signature_def[:args] = unionise_sig.(original_signature_args)
+
+    fdef = quote
+        @inline $(preprocess_declaration(signature_def))
+        @inline $(∇_declaration(signature_def))
+        $(overload_declarations!(signature_def, original_signature_args)...)
+    end
+    #@show fdef
+    eval(fdef)
+    return true
+end
+
+"like `ExprTools.signature` but on a signature type-tuple, not a Method"
+function build_def(sig)
+    def = Dict{Symbol, Any}()
+
+    opT = ExprTools.parameters(sig)[1]
+    def[:name] = :(op::$opT)
+
+    explicit_tvars = Core.TypeName[]#ExprTools.extract_tvars(sig)
+    arg_types = ExprTools.name_of_type.(ExprTools.argument_types(sig))
+    arg_names = [Symbol(:x, ii) for ii in eachindex(arg_types)]  #TODO: should we pass the arg_names in?
+    def[:args] = Expr.(:(::), arg_names, arg_types)
+    def[:whereparams] = ExprTools.where_parameters(sig)
+
+    def = Dict{Symbol, Any}(k => v for (k, v) in def if v !== nothing)  # filter out nonfields.
+
+    return def
+end
+
+"this overwrites and ruins `signature_def` for others"
+function overload_declarations!(signature_def, original_signature_args)
+
+    # Our macro-hygine is not complete here.
+    # the argument names and `op`, `tape` `args`, `kwargs` etc could conflict with
+    # where-params. but for sake of outputting readable code we are not gensyming everything
+    # chance of conflict seems low as where-params are normally upper-case.
+    @assert(signature_def[:name].head == :(::))
+    @assert(signature_def[:name].args[1] == :op)
+    arg_names = [first(arg.args) for arg in signature_def[:args]]
+
+    signature_def[:kwargs] = [:(kwargs...)]
+    signature_def[:body] = quote
+        #@show op
+        args = ($(arg_names...),)
```

Comment:
we need to additionally splat any of the `arg_names` that were themselves Varargs